File: 974.subarray-sums-divisible-by-k.py
#
# @lc app=leetcode id=974 lang=python3
#
# [974] Subarray Sums Divisible by K
#

# @lc code=start
from collections import Counter
class Solution:
    def subarraysDivByK(self, nums: list[int], k: int) -> int:
        count = Counter({0: 1})
        prefix = 0
        ans = 0
        for num in nums:
            prefix = (prefix + num) %k
            ans += count[prefix]
            count[prefix] += 1
        return ans
# @lc code=end

File: 974.subarray-sums-divisible-by-k.java
/*
 * @lc app=leetcode id=974 lang=java
 *
 * [974] Subarray Sums Divisible by K
 */
import java.util.HashMap;
import java.util.Map;
// @lc code=start
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0,1);
        int sum = 0, ans = 0,n = nums.length;
        for(int i = 0; i < n; i++){
            sum += nums[i];
            int mod = (sum % k + k) % k; // to handle negative numbers
            if(map.containsKey(mod)){
                ans += map.get(mod);
            }
            map.put(mod, map.getOrDefault(mod, 0) + 1);
        }
        return ans;
    }
}
// @lc code=end

